### PR TITLE
refactor: Convert 'storage' module to TypeScript

### PR DESCRIPTION
**Pull request by bennyn**

### CODE REVIEW

File: src/script/storage/StorageRepository.ts [hunk 1]
```diff
@@ -34,78 +41,42 @@ export class StorageRepository {
...
-   * Save a value in the amplify value store.
-   *
-   * @param {string} primaryKey - Primary key to save the object with
-   * @param {value} value - Object to be stored
-   * @returns {Promise} Resolves with the primary key
-   */
-  saveValue(primaryKey, value) {
-    return this.storageService.save(this.AMPLIFY_STORE_NAME, primaryKey, {value: value});
+  async saveValue<T>(primaryKey: string, value: T): Promise<string> {
+    await this.storageService.save(this.AMPLIFY_STORE_NAME, primaryKey, {value: value});
```
Comment (ffflorian) [inline src/script/storage/StorageRepository.ts]:
```suggestion
    await this.storageService.save(this.AMPLIFY_STORE_NAME, primaryKey, {value});
```

File: src/script/storage/StorageService.ts [hunk 2]
```diff
@@ -17,25 +17,36 @@
...
 
 import {Config} from '../auth/config';
-import {StorageSchemata} from './StorageSchemata';
-import {StorageKey} from './StorageKey';
 import {ClientType} from '../client/ClientType';
+import {StorageKey} from './StorageKey';
+import {StorageSchemata} from './StorageSchemata';
+
+interface DatabaseListener {
+  callback: Function;
```
Comment (ffflorian) [inline src/script/storage/StorageService.ts]: Can we type this better?

File: src/script/storage/StorageService.ts [hunk 3]
```diff
@@ -68,12 +79,6 @@ export class StorageService {
...
 
-    this.db.on('blocked', event => {
-      if (event.dataLoss !== 'none') {
-        this.logger.error('Database is blocked', event);
-      }
-    });
-
     this._upgradeStores(this.db);
 
     return this.engine
```
Comment (ffflorian) [inline src/script/storage/StorageService.ts]: Why no `async` here?

File: src/script/storage/StorageService.ts [hunk 4]
```diff
@@ -185,32 +205,33 @@ export class StorageService {
...
     this.logger.info(`Clearing object store '${storeName}' in database '${this.dbName}'`);
-    return this.db[storeName].clear();
+    return this.db.table(storeName).clear();
   }
 
   /**
    * Delete multiple database stores.
-   * @param {Array<string>} storeNames - Names of database stores to delete
-   * @returns {Promise} Resolves when the stores have been deleted
+   * @param tring>} storeNames - Names of database stores to delete
```
Comment (ffflorian) [inline src/script/storage/StorageService.ts]:
```suggestion
   * @param storeNames - Names of database stores to delete
```

File: src/script/storage/StorageService.ts (hunk 3 above)
Comment (bennycode) [reply src/script/storage/StorageService.ts]: The JS code was not using `async/await`, so I didn't want to apply it in order to make it easier to compare both versions during the PR review.

Comment (ffflorian) [reply src/script/storage/StorageService.ts]: I see. But in `StorageRepository.ts` you also introduced `async`. I think it's still readable if you convert this one to async :+1:

Comment (bennycode) [reply src/script/storage/StorageService.ts]: Ok, Conversion done with: https://github.com/wireapp/wire-webapp/pull/6887/commits/5a8b6d9d8d61fb3775ae588c112648123c5c96df

File: src/script/storage/StorageService.ts [hunk 5]
```diff
@@ -68,71 +84,79 @@ export class StorageService {
...
+    await this.engine.initWithDb(this.db);
+    await this.db.open();
+
+    try {
+      this._initCrudHooks();
+      this.logger.info(`Storage Service initialized with database '${this.dbName}' version '${this.db.verno}'`);
+      return this.dbName;
+    } catch (error) {
+      const logMessage = `Failed to initialize database '${this.dbName}': ${error.message || error}`;
+      this.logger.error(logMessage, {error: error});
```
Comment (ffflorian) [inline src/script/storage/StorageService.ts]:
```suggestion
      this.logger.error(logMessage, {error});
```

File: src/script/storage/StorageService.ts [hunk 6]
```diff
@@ -68,71 +84,79 @@ export class StorageService {
...
-        this._initCrudHooks();
-        this.logger.info(`Storage Service initialized with database '${this.dbName}' version '${this.db.verno}'`);
-        return this.dbName;
-      })
-      .catch(error => {
-        const logMessage = `Failed to initialize database '${this.dbName}': ${error.message || error}`;
-        this.logger.error(logMessage, {error: error});
-        throw new z.error.StorageError(z.error.StorageError.TYPE.FAILED_TO_OPEN);
-      });
+    await this.engine.initWithDb(this.db);
```
Comment (ffflorian) [inline src/script/storage/StorageService.ts]: Let's move this line into the `try` block, too.

**Pull request merged**